File: C++/ConstructBinaryTreeFromPreorderAndInorderTraversal.cc
// Given preorder and inorder traversal of a tree, construct the binary tree.

// Note:
// You may assume that duplicates do not exist in the tree.

class Solution {
    using position = pair<int, int>;

    int find(vector<int>& v, int target, position& pos) {
        for (int i = pos.first; i < pos.second; i++) {
            if (v[i] == target) {
                return i;
            }
        }
        return -1;
    }

    TreeNode* build(vector<int>& preorder, position&& pp, vector<int>& inorder, position&& ip) {
        if (pp.first == pp.second || pp.second - pp.first != ip.second - ip.first) {
            return nullptr;
        }
        TreeNode* root = new TreeNode(preorder[pp.first]);
        int idx = find(inorder, preorder[pp.first], ip);
        int len = idx - ip.first;
        root->left = build(preorder, {pp.first + 1, pp.first + 1 + len}, inorder, {ip.first, idx});
        root->right = build(preorder, {pp.first + 1 + len, pp.second}, inorder, {idx + 1, ip.second});
        return root;
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder, {0, preorder.size()}, inorder, {0, inorder.size()});
    }
};

File: Java/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
// Given preorder and inorder traversal of a tree, construct the binary tree.

// Note:
// You may assume that duplicates do not exist in the tree.

public class ConstructBinaryTreeFromPreorderAndInorderTraversal {
    TreeNode buildTree(int[] preorder, int l1, int r1, int[] inorder, int l2, int r2) {
        if (l1 > r1) {
            return null;
        }    
        TreeNode root = new TreeNode(preorder[l1]);
        int i = l2;
        while (inorder[i] != preorder[l1]) 
            i++;
        root.left = buildTree(preorder, l1 + 1, l1 + i - l2, inorder, l2, i - 1);
        root.right = buildTree(preorder, l1 + i - l2 + 1, r1, inorder, i + 1, r2);
        return root;
    }
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return buildTree(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);   
    }
}
